frontend/ui/widget/touchmenu.lua

```diff
@@ -961,4 +963,206 @@ function TouchMenu:onBack()
+                table.insert(foundMenuItem, {menu_text, path .. "." .. i, text})
+            end
+        end
+    end
+
+    for i = 1, #self.tab_item_table do
+        recurse(self.tab_item_table[i], i, self.tab_item_table[i].text or "ROOT"..i, 0)
+    end
+
+--[[
+    for i = 1, #foundMenuItem do
+        print("xxxxxxx->", i,
+            foundMenuItem[i][1],
+            foundMenuItem[i][2],
+            foundMenuItem[i][3])
+    end
+]]
+    return foundMenuItem
+end
+
+function TouchMenu:openMenu(path)
+    local animation_time_s = G_reader_settings:readSetting("menu_search_animation_time_s", 1.0)
+    local function animate_func()
+        if not animation_time_s or animation_time_s == 0.0 then return end
+        local sleep_us = animation_time_s * 1e6
+        UIManager:forceRePaint()
+        ffiUtil.usleep(sleep_us)
+        UIManager:setDirty(nil, "ui")
+    end
+    -- first switch to correct MenuTab
+    local sep_pos = path:find("%.")
+    local tab_num = tonumber(path:sub(1, sep_pos - 1))
+    path = path:sub(sep_pos + 1)
+    local item = self.tab_item_table[tab_num]
+
+    if not tab_num then
+        return
+    end
+
+    local old_flash_ui = G_reader_settings:readSetting("flash_ui")
+    G_reader_settings:makeTrue("flash_ui")
+
+    self:switchMenuTab(tab_num)
+    self.bar:switchToTab(tab_num)
+
+    self:updateItems()
+
+    -- Now open the menus all the way down
+    local dummy, num_of_sep = path:gsub("%.", "%.")
+    while num_of_sep > 1 do
+        sep_pos = path:find("%.")
+        local identifier = path:sub(1, sep_pos -1)
+        local item_num = tonumber(identifier)
+        path = path:sub(sep_pos + 1)
+        if item_num then
+            animate_func()
+            item = item[item_num]
+            self:onMenuSelect(item)
+        elseif identifier == "sub_item_table" then
+            item = item.sub_item_table
+        end
+        num_of_sep = num_of_sep - 1
+    end
+
+    -- Now we are in the right menu, but maybe in the wrong page
+    sep_pos = path:find("%.")
+    local identifier = path:sub(1, sep_pos -1)
+    local item_num = tonumber(identifier)
+
+    if item_num then -- might be nil on a sub-menu-entry
+        local desired_page_num = math.floor(item_num / self.perpage)
+        if desired_page_num > 0 then
+            for i = 1, desired_page_num do
+                animate_func()
+                self:onNextPage()
+            end
+        end
+    end
+
+    G_reader_settings:saveSetting("flash_ui", old_flash_ui)
+end
+
+function TouchMenu:onMenuSearchDialog()
+    local InputDialog = require("ui/widget/inputdialog")
+    local CheckButton = require("ui/widget/checkbutton")
+    local ConfirmBox = require("ui/widget/confirmbox")
+
+    local function show_search_results(search_string)
+        local foundMenuItem = self:search(search_string)
+
+        local function get_current_search_results()
+            local function item_callback(i)
```

Comment:
(That's a lot of nested functions  :)- but I'm fine with it as it makes it all self contained in these methods.)

Reply:
(Nesting is not that performant, but that was the idea.)